### ISSUE TITLE
On the future of this project

### ISSUE THREAD

**Issue by mserajnik** [mserajnik/hydrusrv#18]
I've been working on hydrusrv and its "companion applications" for the better part of a year now. What started as a small side project with a very limited scope grew into a decent-sized application with quite extensive search and sort options, more elaborate authentication than I ever intended and a good amount of options to personalize all of it.

I'm very happy on how it turned out, even if it took much longer than expected to reach the point it is at now which I'd consider "feature complete" (although that's also due to the scope always increasing). There's not much else I could add in terms of searching/sorting without deviating from the hydrus server data structure and starting to add my own computed data on top of it (which would defeat the whole purpose).

It's certainly a nice thing to know that you've pretty much exhausted the possibilities of what you can add (better: of things that make sense to add).

That said, not everything is great, though.

The biggest thing that bugs me is the naming. When I started working on it, _hydrusrv_ seemed like a clever acronym for _hydrus_ and _server_, with _hydrusrvue_ fitting as well, since I just had to add `ue` for it to describe the main technology used in the frontend (Vue).

However, now that hydrusrv no longer accesses the hydrus server databases directly and instead relies on another tool (hydrusrv-sync) to make a transformed copy of the data, I feel like the name is very limiting; one could easily write a tool that would "sync" the data from a different source than hydrus server. E.g., something more along the lines of what [Plex](http://plex.tv) does, reading files from a directory and building its database this way.

Another issue I have is with performance, more specifically requests per second. While the change to run multiple workers did help a lot, scaling hydrusrv to several thousand RPS is still not trivial and requires quite a bit of hardware. And even if it was never my intention for someone to run hydrusrv on a large scale like Danbooru or other platforms, I at least don't want there to be such technical limitations that it couldn't compete against other applications if it came down to a "price to run it" comparison.

There are more things that I'm not totally happy about, but in the interest of not writing a novel:

Ultimately, I want hydrusrv to stand on its own. I've therefore decided to stop new feature development on hydrusrv and its associated projects:

- [hydrusrvue](https://github.com/mserajnik/hydrusrvue)
- [hydrusrv-sync](https://github.com/mserajnik/hydrusrv-sync)
- [hydrusrv-docker](https://github.com/mserajnik/hydrusrv-docker)

I will still actively maintain [hydrus-server-docker](https://hub.docker.com/r/mserajnik/hydrus-server-docker) as it can be used standalone and I have need for it anyway (more on that further down).

"No new features" means that I'll limit any future development on pure maintenance tasks (bug fixes, dependency updates etc.).

Meanwhile, I'll start working on a successor to hydrusrv and the other applications. Something that works in a similar way, but with a new name and without having direct ties to hydrus server. I've not yet decided on all the details, but I'll likely do the following:

- Rewrite hydrusrv as this new application, keeping the same feature set and/or expanding it and making adjustments where it makes sense; this project might be written with Node/Express again or use a different tech stack (I'm currently trying to invest more time into learning other languages like [Go](https://golang.org/) and [Rust](https://www.rust-lang.org/))
- Rewrite hydrusrvue as a frontend for the new application once Vue 3.0 comes out, maybe without Bulma and more custom styling instead. It might also get additional features such as connecting to multiple servers and switching on the fly instead of being tied to a single one
- Rewrite hydrusrv-sync as the first in a line of tools that I call "data providers" – basically a tool that fills the new application's database. This one will once again get its data from hydrus server (so all of this _can_ still work just like the current applications do). Again, might be in Node, might be in a different language that makes sense
- Provide Docker images and a Docker Compose setup, just like I did with hydrusrv-docker

To sum up, the new applications will be __a full replacement for hydrusrv and associated applications__, just better. More features, better performance, better usability – I'll try to get as much feedback during development as I can to make it feel like an upgrade that's worth it.

I have no exact timeframe of when all of this will be done, but the current plan is to release `1.0` versions some time in the second half of this year. I'll update this thread whenever I have more information.

Once the new applications are released, I'll cease development on hydrusrv and the other applications altogether and archive the repositories.

Finally, thanks for all the help, feedback and contributions – you know who you are.

**Comment by mserajnik** [comment]
The new project is called _Hydrarch_ and lives under its own organization: https://github.com/hydrarch

Target release date is Q3 2019.
